### ISSUE TITLE
Improvement: Formatting rule descriptions with the record

### ISSUE THREAD

**Issue by stoggi** [airbnb/streamalert#464]
## Background

cc: @jacknagz @ryandeivert @0xdabbad00

I propose a simple way of formatting slack messages requirement in #416.

>Ability to include items from the record into the alert and to enhance that data. We might want our alerts to just include "Root login to AWS account Prod (111111111111)". This would mean no description, no record listing, just that line of text. That line of text would extract out the relevant account ID, and would enhance it by doing a lookup that gives the human readable name of that account.

The rule description could be formatted using Python's standard `string.format()` function, with a simple change to the SlackOutput output class:

```python
@classmethod
def _format_message(cls, rule_name, alert):
    ...
    try:
        rule_desc = alert['rule_description'].format(**alert['record'])
    except (IndexError, KeyError):
        rule_desc = alert['rule_description']
    ...
```

This would allow you to write a rule description like this:

```python
def cloudtrail_assume_role_organization_account_access(rec):
    """User `{userIdentity[userName]}` has assumed *OrganizationAccountAccessRole* from `{sourceIPAddress}`.

    description:      This role should only be assumed when performing root level actions in an account.
    reference:        https://reference.com
    playbook:         (a) determine the account from the arn `{requestParameters[roleArn]}`.
                      (b) ping `{userIdentity[userName]}` to determine if intentional and/or legitimate
    """
```

This would generate a slack message that looks similar to:

>User `bob.smith` has assumed **OrganizationAccountAccessRole** from `54.32.12.34`.
>
>    description:      This role should only be assumed when performing root level actions in an account.
>    reference:        https://reference.com
>    playbook:         (a) determine the account from the arn `arn:aws:iam::112233445566:role/Role`.
>                      (b) ping `bob.smith` to determine if intentional and/or legitimate

You could even remove the description and playbook from the docstring on certain rules:

>User `bob.smith` has assumed **OrganizationAccountAccessRole** from `54.32.12.34`.

## Desired Change

A simple change to the outputs can allow fields in the record dictionary to be used in the rule description.

I would also put the rule_description into the main text body of the slack message, so that when you share it, it includes the formatted content. Attachments don't appear on the shared message.

**Comment by 0xdabbad00** [comment]
This would be an improvement over the current master, but would not be sufficient for all of the needs I had that required me to create #416.

Additional requirements I had:
- Not display the entire docstrings anymore.  That could be an easy change to have a flag to display the docstring or not.
- Display a human readable name for accounts along with the 12-digit AWS account ID.
- Be able to perform some routing of the alerts based on the account ID.

**Comment by stoggi** [comment]
Hmm. That is a good point. I didn't mean to suggest this covered all the requirements, it was just a suggestion for formatting the docstring.

It's almost like you need a third lambda function to format alerts that triggers for one or more matched rules and one or more outputs. It could be passed the rule description and the record.

That way the alert processor just handles how to alert to a particular output, not necessarily what to alert. You could then put more complicated lookups in that formatting processor. 

If no formatting processor is defined then it could use a default one.

**Comment by stoggi** [comment]
It would be great to have a short formatted message in one slack channel, and a more comprehensive enriched message in another slack channel and s3 for example.

**Comment by Ryxias** [comment]
I think this has been addressed through [AlertPublishers](https://streamalert.io/en/stable/publishers.html), released in [2.2.0](https://github.com/airbnb/streamalert/releases/tag/2.2.0).
